### PR TITLE
Support heterogeneous DDP overlap in multimodal model training

### PR DESCRIPTION
**Pull request by yashaswikarnati**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

**Allows multimodal  models to use DDP-wrapped submodules with different
parameter-gather overlap policies while participating in the standard
Megatron training, checkpoint, rerun, and paged-stash lifecycles.**

## Issue tracking

Linked issue: N/A — compatibility gaps found during MIMO training integration.

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code
- [ ] I have added relevant documentation
- [ ] I have run `tools/autoformat.sh` on my PR

## Validation

- CMH job `3188397`: 20 focused tests passed; 3 expected 8-GPU tests skipped (1 node / 4 allocated GPUs / 1 pytest process).
- Black 26.3.0, isort 5.13.2, Ruff 0.9.10, Pylint 3.2.6, Python compilation, and `git diff --check` pass on the changed scope. `training.py` imports are unchanged and retain pre-existing full-file isort debt.

### CODE REVIEW

**Review by cuichenx — APPROVED** [review]

**Review by fanshiqing — APPROVED** [review]

File: examples/mimo/training/builder.py [hunk 1]
```diff
@@ -24,7 +24,7 @@
 from megatron.training.global_vars import get_args
 from megatron.training.models.base import ModelBuilder, ModelConfig, compose_hooks
 
-_LANGUAGE_SEED_OFFSET = 20_000
```
Comment (deepakn94) [inline examples/mimo/training/builder.py:27]: What is this?

File: megatron/training/training.py [hunk 2]
```diff
@@ -3058,7 +3058,11 @@ def train_step(forward_step_func, data_iterator, model, optimizer, opt_param_sch
             full_cg_captured = FullCudaGraphWrapper.cuda_graph.get("training") is not None
             if forward_pre_hook_enabled or full_cg_captured:
                 for optim_instance in optimizer.chained_optimizers:
-                    if isinstance(optim_instance, DistributedOptimizer):
+                    if (
```
Comment (deepakn94) [inline megatron/training/training.py]: Add comment on why this changed?

File: examples/mimo/training/builder.py [hunk 3]
```diff
@@ -24,7 +24,7 @@
...
+_LANGUAGE_SEED_OFFSET = 0
```
Comment (yashaswikarnati) [inline examples/mimo/training/builder.py:27]: this is the seed passed to _set_random_seed(),and which derives the data/tensor/expert parallel seeds for the rng tracker. I changed this to match pretrain_hrybrid, which passes seed 0. and the actual historic reason for having differnt encoder and language model seed offset is that in mimo encoders and LLM can be colocated or non colocated with different paralleisms

File: megatron/training/training.py
Comment (yashaswikarnati) [inline megatron/training/training.py]: the outer loop uses args to check for overlap settings which are global, but we want to have encoders and LLM independently choose different overlap settings. the below checks uses the module's optimizer instance's ddp config to indepdently derive the overlap and reuse grad buffer settings

**Review by deepakn94 — APPROVED** [review]

**Comment by yashaswikarnati** [comment]
/ok to test [1670ebb](https://github.com/NVIDIA/Megatron-LM/pull/6284/commits/1670ebbca9002030c047045a7f49514ad4648afc)

**Comment by yashaswikarnati** [comment]
/ok to test  [de0ca4e](https://github.com/NVIDIA/Megatron-LM/pull/6284/commits/de0ca4ebca898b72bb51d8282e21c06e3c152d7c)

**Comment by yashaswikarnati** [comment]
/ok to test [de0ca4e](https://github.com/NVIDIA/Megatron-LM/pull/6284/commits/de0ca4ebca898b72bb51d8282e21c06e3c152d7c)
